outlinecaddy/websocket2layer4_handler.go

```diff
@@ -0,0 +1,167 @@
+	// not provided, it defaults to StreamConnectionType.
+	Type ConnectionType `json:"type,omitempty"`
+
+	// ConnectionHandler specifies the name of the connection handler to use. This
+	// name must match a handler configured within the Outline app.
+	ConnectionHandler string `json:"connection_handler,omitempty"`
+
+	// compiledHandler is the compiled instance of the named connection
+	// handler. It is populated during the Provision step.
+	compiledHandler layer4.NextHandler
+
+	logger  *slog.Logger
+	zlogger *zap.Logger
+}
+
+var (
+	_ caddy.Provisioner           = (*WebSocketHandler)(nil)
+	_ caddy.Validator             = (*WebSocketHandler)(nil)
+	_ caddyhttp.MiddlewareHandler = (*WebSocketHandler)(nil)
+)
+
+func (*WebSocketHandler) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{ID: wsModuleName}
+}
+
+// Provision implements caddy.Provisioner.
+func (h *WebSocketHandler) Provision(ctx caddy.Context) error {
+	h.logger = ctx.Slogger()
+	h.zlogger = ctx.Logger()
+	if h.Type == "" {
+		// The default connection type if not provided is a stream.
+		h.Type = StreamConnectionType
+	}
+
+	mod, err := ctx.AppIfConfigured(outlineModuleName)
+	if err != nil {
+		return fmt.Errorf("outline app configure error: %w", err)
+	}
+	app, ok := mod.(*OutlineApp)
+	if !ok {
+		return fmt.Errorf("module `%s` is of type `%T`, expected `OutlineApp`", outlineModuleName, app)
+	}
+	for _, compiledHandler := range app.Handlers {
+		if compiledHandler.Name == h.ConnectionHandler {
+			h.compiledHandler = compiledHandler
+			break
+		}
+	}
+	if h.compiledHandler == nil {
+		return fmt.Errorf("no connection handler configured for `%s`", h.ConnectionHandler)
+	}
+
+	return nil
+}
+
+// Validate implements caddy.Validator.
+func (h *WebSocketHandler) Validate() error {
+	if h.Type != "" && h.Type != StreamConnectionType && h.Type != PacketConnectionType {
+		return fmt.Errorf("unsupported `type`: %v", h.Type)
+	}
+	if h.ConnectionHandler == "" {
+		return errors.New("must specify `connection_handler`")
+	}
+
+	return nil
+}
+
+// ServeHTTP implements caddyhttp.MiddlewareHandler.
+func (h WebSocketHandler) ServeHTTP(w http.ResponseWriter, r *http.Request, _ caddyhttp.Handler) error {
+	h.logger.Debug("handling connection",
+		slog.String("path", r.URL.Path),
+		slog.Any("remote_addr", r.RemoteAddr))
+
+	conn, err := websocket.Upgrade(w, r, nil)
+	if err != nil {
+		h.logger.Error("failed to upgrade", "err", err)
+	}
+	defer conn.Close()
+	if clientIpStr, ok := caddyhttp.GetVar(r.Context(), caddyhttp.ClientIPVarKey).(string); ok {
+		if clientIp := net.ParseIP(clientIpStr); clientIp != nil {
+			switch h.Type {
+			case StreamConnectionType:
+				conn = &replaceAddrConn{StreamConn: conn, raddr: &net.TCPAddr{IP: clientIp}}
+			case PacketConnectionType:
+				conn = &replaceAddrConn{StreamConn: conn, raddr: &net.UDPAddr{IP: clientIp}}
+			}
+		}
+	}
+	cx := layer4.WrapConnection(conn, []byte{}, h.zlogger)
```

Comment:
WrapConnection seems to potentially add significant overhead with the rewind logic. It will be helpful to understand the impact later, and we can consider moving away from the layer4 module.